### ISSUE TITLE
Warn if rendering an HTMLUnknownElement

### ISSUE THREAD

**Issue by spicyj** [facebook/react#3726]
This usually means that the user mixed up our lowercase/uppercase naming convention; see https://github.com/facebook/react/pull/3429#issuecomment-85166440.

We could add a message like

> Warning: The tag <commentBox> is unrecognized in this browser. If you meant to render a React component, start its name with an uppercase letter.

We'd check this only once per tag and only in development mode.

**Comment by clentfort** [comment]
For a good first bug a note on where to start reading into the code would be helpful. I took a look at [ReactElementValidator.js](https://github.com/facebook/react/blob/master/src/classic/element/ReactElementValidator.js) but am not sure it is the right file to add this functionality to.

**Comment by spicyj** [comment]
I'd probably add it to the logic in validateDangerousTag inside ReactDOMComponent.js, as that already runs only once per tag. Note that that runs in dev and prod mode for safety, but this new check should probably be wrapped in an `if (__DEV__)`.

**Comment by bloodyowl** [comment]
actually tried to implement this in #3843, could be more complicated than I originally thought as svg (and descendant) elements are instances of `HTMLUnknownElement`, which would require a whitelist. not sure if it's a good idea.

**Comment by kastermester** [comment]
@spicyj Do you have any idea how to handle svg elements and the likes?

I tried doing something like:

```javascript
if (__DEV__) {
  var isHtmlElement = !(document.createElement(tag) instanceof HTMLUnknownElement);
  var isSvgElement = !(document.createElementNS('http://www.w3.org/2000/svg', tag) instanceof HTMLUnknownElement);
  warning(isHtmlElement || isSvgElement, 'Some warning here');
}
```

This works for me when I run the code in my chrome debugger, however when running `grunt test` it doesn't.

There's a couple of other edge cases as well (noscript tags, web components) that breaks the tests as well. Which I haven't quite gotten around to fixing other than simple ugly fixes.

**Comment by spicyj** [comment]
Hmm, SVG makes this a little trickier. Ideally we'd know whether we're in an SVG container or not (which we need for a handful of other stuff) – but we don't right now so this may have to wait on that. Thanks for looking into this @kastermester.

**Comment by gaearon** [comment]
>Hmm, SVG makes this a little trickier. Ideally we'd know whether we're in an SVG container or not (which we need for a handful of other stuff) – but we don't right now so this may have to wait on that. 

We do have this information now in `ReactDOMComponent`. Retagging.

**Comment by reznord** [comment]
@spicyj @gaearon I would like to take up this bug and fix it.

**Comment by gaearon** [comment]
@reznord Cool, please give it a try!

**Comment by reznord** [comment]
@gaearon tests are failing when I try running **npm run test**. Can you tell me where I am going wrong?

log: https://gist.github.com/reznord/5dba51c3449e03c98176b166e4ae68f7

**Comment by ConnorAtherton** [comment]
@reznord I did a little digging around this and want to add a few comment in case you are still interested in this feature. If not, I have some code running already that I can PR.

The code example given in https://github.com/facebook/react/pull/3429#issuecomment-85166440 works correctly for detecting valid element in the default namespace. SVG, however, doesn't play nicely. Using the same check with the svg namespace doesn't work as expected. (I've used `HTMLUnknownElement` here since I was testing on a modern version of Chrome, Firefox, and Safari. This would be replaced to also support `HTMLGenericElement` for old browsers. )

```
document.createElementNS('http://www.w3.org/2000/svg', 'thisIsNotAValidElement') instanceof HTMLUnknownElement
=> false
```
Searching for some answers on how to correcty handle this, I checked the docs to no avail, and tried to scan through the d3 source to see if you how are handling invalid svg elements in any meaningful way but found nothing https://github.com/d3/d3-selection/blob/master/src/creator.js#L10. 

@gaearon What information do you think would be useful for this? I can see that `_namespaceURI` is tracked as a property on `ReactDOMComponent`, mainly for `mountComponent`, but knowing what DOMNamespace to test against doesn't help us since the condition will still return false. 

Another closely related point to this would be the question of how to attempt to detect correct [custom elements](https://www.w3.org/TR/custom-elements/). Or, even if we should even try to detect them?

Since this is an additional feature enabled only in development to help developers spot tiny mistakes -- typo, like <spann>, for example -- can we handle the most common case and add the check using the default namespace with `document.createElement` for now?

If I have overlooked something, please let me know. I've never attempted to verify valid svg tags in the browser and I'm very interested to know if there is a stable solution to this :)

**Comment by gaearon** [comment]
I think it's fine if we don't warn inside SVG as long as we don't have false positives.

**Comment by reznord** [comment]
@ConnorAtherton I was working on the same and made some progress. I will be sending a PR soon to this. Feel free to send a PR if you have already worked on it. :)

**Comment by arshabh** [comment]
@gaearon I have added the warning check in `transaction.useCreateElement`. Is that fine ?

**Comment by JonDum** [comment]
> This usually means that the user mixed up our lowercase/uppercase naming convention;

Now that React 16 is out, I've experimented upgrading on a few of our apps. There are now several warnings for `HTMLUnknownElement` elements. While the warning don't really affect anything other than making our developers question their sanity and if something is broken, I'd just like to chime in that there are valid use cases for `HTMLUnknownElement`. We use them for semantics to avoid "div hell", e.g. `<spacer/>` to add a flexible gap in flex boxes, `<icon/>` for simple icon fonts that don't need to be react components, etc, etc.

**Comment by Wildhoney** [comment]
I get the warning when using the [`<sup />` tag](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/sup). Is that intended?

> Warning: The tag <sup> is unrecognized in this browser. If you meant to render a React component, start its name with an uppercase letter.

**Comment by gaearon** [comment]
@Wildhoney That doesn't seem right. Also doesn't seem to happen for me: https://jsfiddle.net/fzuo1cLj/

Can you create an isolated example that triggers this? And preferably file a new issue since it's easy to lose reports in old issues.

**Comment by Wildhoney** [comment]
Sorry @gaearon &ndash; I was probably a bit hasty in posting that message. I get the warning when running tests via Enzyme only, so perhaps it's related to that. I'll investigate and take further action if necessary 👍

**Comment by gaearon** [comment]
Oh. That means it’s unsupported in jsdom. Mind filing a separate issue about this? Maybe we can somehow disable it in jsdom environment.

**Comment by AntiFish03** [comment]
@JonDum Agreed and I also use them for things like `<author itemProp="reviewAuthor">` that Search Engines like (since most of them now load and run JS).  But with React 16 I am getting
```
Warning: The tag <author> is unrecognized in this browser. If you meant to render a React component, start its name with an uppercase letter.
```
And yes it does make me question my sanity...
